osu.Game.Rulesets.Osu/UI/Cursor/CursorRippleVisualiser.cs

```diff
@@ -0,0 +1,100 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Pooling;
+using osu.Framework.Input.Bindings;
+using osu.Framework.Input.Events;
+using osu.Game.Rulesets.Osu.Configuration;
+using osu.Game.Rulesets.Osu.Skinning.Default;
+using osu.Game.Skinning;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.UI.Cursor
+{
+    public partial class CursorRippleVisualiser : CompositeDrawable, IKeyBindingHandler<OsuAction>
+    {
+        private readonly Bindable<bool> showRipples = new Bindable<bool>(true);
+
+        private readonly DrawablePool<CursorRipple> ripplePool = new DrawablePool<CursorRipple>(20);
+
+        public CursorRippleVisualiser()
+        {
+            RelativeSizeAxes = Axes.Both;
+        }
+
+        [BackgroundDependencyLoader(true)]
+        private void load(OsuRulesetConfigManager? rulesetConfig)
+        {
+            rulesetConfig?.BindWith(OsuRulesetSetting.ShowCursorRipples, showRipples);
+        }
+
+        public bool OnPressed(KeyBindingPressEvent<OsuAction> e)
+        {
+            if (showRipples.Value)
+            {
+                AddInternal(ripplePool.Get(r =>
+                {
+                    r.Position = e.MousePosition;
+                }));
+            }
+
+            return false;
+        }
+
+        public void OnReleased(KeyBindingReleaseEvent<OsuAction> e)
+        {
+        }
+
+        private partial class CursorRipple : PoolableDrawable
+        {
+            [BackgroundDependencyLoader]
+            private void load()
+            {
+                AutoSizeAxes = Axes.Both;
+                Origin = Anchor.Centre;
+
+                InternalChildren = new Drawable[]
+                {
+                    new SkinnableDrawable(new OsuSkinComponentLookup(OsuSkinComponents.CursorRipple), _ => new DefaultCursorRipple())
+                    {
+                        Blending = BlendingParameters.Additive,
+                    }
+                };
+            }
+
+            protected override void PrepareForUse()
+            {
+                base.PrepareForUse();
+
+                ClearTransforms(true);
+
+                this.ScaleTo(0.05f)
+                    .ScaleTo(1, 700, Easing.Out)
+                    .FadeOutFromOne(700)
+                    .Expire(true);
+            }
+        }
+
+        public partial class DefaultCursorRipple : CompositeDrawable
+        {
+            [BackgroundDependencyLoader]
+            private void load()
+            {
+                AutoSizeAxes = Axes.Both;
+
+                InternalChildren = new Drawable[]
+                {
+                    new RingPiece(3)
+                    {
+                        Size = new Vector2(256),
+                        Alpha = 0.2f,
```

Comment:
So the above comment has the right sentiment but incorrect execution. The disparity between lazer and stable comes from two things.

Firstly, it seems that the 256px size spec in stable is in *real* screen pixels. (It's likely because in stable the ripples work everywhere, not just in gameplay.) Currently in lazer, because the ripples are implemented locally in `OsuCursorContainer`, the 256 size spec is in "osu!pixels" or whatever else you want to call the playfield units. That's why the ripples are larger - but in general it is incorrect to hardcode another value, because the scaling factor to compensate by depends on the user's resolution. The "proper" solution *could* look something like:

```diff
diff --git a/osu.Game.Rulesets.Osu/UI/Cursor/CursorRippleVisualiser.cs b/osu.Game.Rulesets.Osu/UI/Cursor/CursorRippleVisualiser.cs
index 465b708b5a..7ab2670dae 100644
--- a/osu.Game.Rulesets.Osu/UI/Cursor/CursorRippleVisualiser.cs
+++ b/osu.Game.Rulesets.Osu/UI/Cursor/CursorRippleVisualiser.cs
@@ -39,6 +39,7 @@ public bool OnPressed(KeyBindingPressEvent<OsuAction> e)
                 AddInternal(ripplePool.Get(r =>
                 {
                     r.Position = e.MousePosition;
+                    r.Scale = new Vector2(OsuPlayfield.BASE_SIZE.X / ScreenSpaceDrawQuad.Width);
                 }));
             }
 
@@ -51,6 +52,8 @@ public void OnReleased(KeyBindingReleaseEvent<OsuAction> e)
 
         private partial class CursorRipple : PoolableDrawable
         {
+            private Container content = null!;
+
             [BackgroundDependencyLoader]
             private void load()
             {
@@ -59,9 +62,12 @@ private void load()
 
                 InternalChildren = new Drawable[]
                 {
-                    new SkinnableDrawable(new OsuSkinComponentLookup(OsuSkinComponents.CursorRipple), _ => new DefaultCursorRipple())
+                    content = new Container
                     {
-                        Blending = BlendingParameters.Additive,
+                        new SkinnableDrawable(new OsuSkinComponentLookup(OsuSkinComponents.CursorRipple), _ => new DefaultCursorRipple())
+                        {
+                            Blending = BlendingParameters.Additive,
+                        }
                     }
                 };
             }
@@ -72,9 +78,10 @@ protected override void PrepareForUse()
 
                 ClearTransforms(true);
 
-                this.ScaleTo(0.05f)
-                    .ScaleTo(1, 700, Easing.Out)
-                    .FadeOutFromOne(700)
+                content.ScaleTo(0.05f)
+                       .ScaleTo(1, 700, Easing.Out);
+
+                this.FadeOutFromOne(700)
                     .Expire(true);
             }
         }
```

Secondly, the alpha disparity is due to the fact that osu!stable - if a `cursor-ripple` asset is not present in the current skin - instead uses a built-in `menu-osu-shockwave` UI asset. That asset is a gradiented circle, which at the outer edges has about 40% alpha:

![image](https://user-images.githubusercontent.com/20418176/235372155-6319f1a9-1eec-4110-b03a-436d8574399e.png)

Note that 0.2 * 0.4 = 0.08, so that tracks if you wanted to make a solid ring (which lazer is currently using as the default implementation) match. That however means that the 0.2 alpha spec is technically *correct*, it's just that the default implementation doesn't match for the legacy default skin.

Reply:
> in general it is incorrect to hardcode another value, because the scaling factor to compensate by depends on the user's resolution

The idea was to implement this local to gameplay and make the size match that usage better, but I guess there is argument for it being implemented game-wide. Let's ignore that for now, though.

I think we want this element to match gameplay sizing and potentially scale with CS (as has been requested) as long as it is gameplay local, so a new constant (similar to proposed) is probably the best path forward?

Reply:
As for the alpha, I was aware of the difference. I think it works quite well at `0.2f` but I'll drop it down to `0.1f` for now and we can see what the response is like.

Reply:
In that case I suppose changing the constant size to another value is best. We'll need to communicate the fact that the implementation won't always match 1:1 on all resolutions, however.

From my math, the new constant could be equal to something along the lines of 260×512÷1080=123.25925925925, to match at 1080p (which should be the most common resolution). I haven't visually tested this, though, so the slightly smaller one proposed above may be better.

Reply:
> As for the alpha, I was aware of the difference. I think it works quite well at 0.2f but I'll drop it down to 0.1f for now and we can see what the response is like.

Or you could apply 0.4f alpha to the default implementation for the same result. If you change to 0.1f then user skins that have `cursor-ripple` won't technically match.

Reply:
I brought this up with the team, and as per my inline comments I kinda want to break this intentionally as it's hardly documented on the wiki (no mention of size of alpha). Willing to revisit in the future, just want to get this in for those users which like it. I've never seen it actually skinned outside of one forum thread where someone accidentally found out it's skinnable and was experimenting with it (https://osu.ppy.sh/community/forums/topics/1098674?n=1).